### PR TITLE
Winrm improvements

### PR DESCRIPTION
**Pull request by smashery**
This turns the WinRM command runner into a full-fledged command shell with WinRM as the transport. A bunch of things to neaten up yet, but basic functionality is there; happy for early feedback on my approach to it.

Some design notes:
- I use WinRM's stdin mechanism, rather than the command-based one. This is to support situations such as nested shells (Ruby's WinRM module's default Powershell shell seems to immediately terminate any processes, and its default Cmd shell waits for the process to end, thus seeming to hang when you launch cmd.exe for example, neither of which is fit for our purpose)
- To support streaming output back to Metasploit, I'll probably need to have it regularly requesting data (that's yet to do). Currently it only requests stdout/stderr after sending new stdin.

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] `use scanner/winrm/winrm_cmd`
- [ ] `set rhosts ...`
- [ ] `set username ...`
- [ ] `set password ...`
- [ ] `run`
- [ ] `sessions #`

Test cases to then look at:

- Module:
  - Single command only (`set CreateSession false`)
  - Valid creds
  - Invalid creds
- CreateSession
  - Valid
  - Invalid
- Shell:
  - Nested shells
  - Gracefully handle end of shell:
    - Terminating by user (exit)
    - Terminating the powershell process itself (kill proc)
    - System ceases responding
  - Ensure it's pivot-friendly
  - Long-running program (where output isn't all immediate - should come in as it )

### CODE REVIEW

**Comment by bcoles** [comment]
Related #11456
Possibly related #8900

**Comment by smashery** [comment]
So I believe this is ready for a proper review. As an added bonus, it now supports encryption when using NTLMSSP auth, which should enable it to be used on default installs of Win10 and Server 2019.

Here are the test cases I think are relevant, but let me know if you think of any other edge cases. The focus of this PR is `winrm_login`, but I also ported some changes to `winrm_cmd` to support encryption.

- [ ] `winrm_cmd`:
  - [ ] Valid creds
  - [ ] Invalid creds
- [ ] `winrm_login`:
  - [ ] Valid creds
  - [ ] Invalid creds
- [ ] Working over SSL
- [ ] Nested shells (e.g. powershell launching cmd launching powershell...)
- [ ] Gracefully handle:
  - [ ] Terminating by user (`exit`)
  - [ ] Terminating the powershell process itself (kill the powershell process in task manager)
  - [ ] System ceases responding (e.g. pause your VM)
  - [ ] All of these termination cases while interacting and not interacting with the shell
- [ ] Long-running program (with output that doesn't come all-at-once, but over several minutes) e.g. Seatbelt

I don't expect it to work if NTLM is disabled (i.e. Kerberos-only). I'm not sure about the impact of EPA.

**Comment by smashery** [comment]
Some notes on implementation:

- To ensure we're using Rex sockets, we override a few classes from the WinRM module to force them to use the Rex HttpClient.
- To support stdin and nested shells, we just launch powershell.exe and operate over stdin and stdout, rather than using separate commands. This is different to how tools such as evil-winrm work, which are more command-based. There are pros and cons both ways; this was chosen as it's closer to how other MSF command shells work.
- A background thread runs constantly to ask "is there more stdout for me?". This is to be alerted to lost shells while not interacting with them.

File: lib/net/winrm/connection.rb [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+            attr_accessor :http_client, :uri, :timeout
+
+            # Need to send an empty first request to potentially set up an encryption
+            # channel - required if allowUnencrypted is set to false, which is the case
+            # by default
+            attr_accessor :first_request_sent
+        end
+    
+        def create_transport(connection_opts)
+          raise NotImplementedError unless connection_opts[:transport] == :rex
```
Comment (smcintyre-r7) [inline lib/net/winrm/connection.rb:248]: Could we please rename our transport to `:rex_http`? That would more closely align with the `RexHttpTransport` class name and put us in a better position if we got around to adding SSH as a transport.

File: lib/msf/base/sessions/winrm_command_shell.rb [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+      framework.events.on_session_interact_completed()
+      framework.sessions.deregister(self, reason)
+    end
+
+    def stop_keep_alive_loop
+      self.keep_alive_thread.kill
+    end
+
+    def tunnel_to_s
+      'WinRM'
```
Comment (smcintyre-r7) [inline lib/msf/base/sessions/winrm_command_shell.rb:142]: Could you please add the peer information in here? This typically has the local host/port information and the remote host/port information.

If `#tunnel_local` and `#tunnel_peer` are defined, then the `tunnel_to_s` will take over from within [`lib/msf/core/session.rb`](https://github.com/rapid7/metasploit-framework/blob/a7d255bbe5537822c614ede71933fdc6597dd369/lib/msf/core/session.rb#L135). The local port may change if a new socket is opened for each HTTP request but at least having the local IP, remote IP and port would be really helpful.

File: lib/msf/core/exploit/remote/winrm.rb [hunk 3]
```diff
@@ -59,7 +59,7 @@ def winrm_run_cmd(cmd, timeout=20)
       return retval
     end
     shell_id = winrm_get_shell_id(resp)
-    resp = send_winrm_request(winrm_cmd_msg(cmd, shell_id),timeout)
+    resp = eend_winrm_request(winrm_cmd_msg(cmd, shell_id),timeout)
```
Comment (smcintyre-r7) [inline lib/msf/core/exploit/remote/winrm.rb:62]: I don't see where `eend_winrm_request` is defined. Is this a typo?

File: lib/msf/core/exploit/remote/winrm.rb [hunk 4]
```diff
@@ -233,6 +233,9 @@ def send_winrm_request(data, timeout=20)
       'password' => datastore['PASSWORD'],
       'ctype' => "application/soap+xml;charset=UTF-8"
     }
+    print_good(opts['username'])
+    print_good(opts['password'])
+    print_good(opts['uri'])
```
Comment (smcintyre-r7) [inline lib/msf/core/exploit/remote/winrm.rb:238]: If we're keeping these lines could you please add some context around them so users know what values they're looking at?

File: lib/net/winrm/connection.rb [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+          # @yieldparam [string] standard out response text
+          # @yieldparam [string] standard error response text
+          # @yieldreturn [WinRM::Output] The command output
+          def read_stdout(&block)
+            open unless shell_id
+            create_proc unless command_id
+            begin
+              response_reader.read_output(command_output_message(shell_id, command_id), &block)
+            rescue WinRM::WinRMWSManFault => err
+              if err.fault_code == '2150858793'
```
Comment (smcintyre-r7) [inline lib/net/winrm/connection.rb:107]: A comment about what this value is would be really helpful.

File: lib/msf/base/sessions/winrm_command_shell.rb [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+require 'msf/util/signallable_event'
+require 'winrm'
+
+module Msf::Sessions
+  #
+  # This class provides a session for WinRM client connections, where Metasploit
+  # has authenticated to a remote WinRM instance.
+  #
+  class WinrmCommandShell < Msf::Sessions::CommandShell
```
Comment (smcintyre-r7) [inline lib/msf/base/sessions/winrm_command_shell.rb:11]: So if we're actually running Powershell, I think we should be inheriting from `Msf::Sessions::Powershell`. While Powershell also inherits from `CommandShell`, the important bit is that it will correctly set the session type to `"powershell"` which is important for our post libraries.

File: lib/msf/core/session_manager.rb [hunk 7]
```diff
@@ -223,6 +223,10 @@ def register(session)
       end
     end
 
+    if session.respond_to?("on_registered")
```
Comment (adfoster-r7) [inline lib/msf/core/session_manager.rb:226]: What are your thoughts on moving this into the `if session.register?` branch? 👀

File: lib/rex/proto/http/client.rb [hunk 8]
```diff
@@ -502,23 +511,32 @@ def negotiate_auth(opts={})
...
+      self.ntlm_client = ntlm_client
       return resp
 
     rescue ::Errno::EPIPE, ::Timeout::Error
       return nil
     end
   end
 
-  #
+  def channel_binding()
```
Comment (adfoster-r7) [inline lib/rex/proto/http/client.rb:532]:
```suggestion
  def channel_binding
```

**Comment by adfoster-r7** [comment]
Awesome work! Just ran through a few quick scenarios 🕵️ 

### windows server 2019 Domain controller

Worked great, interacting with a session worked as expected :+1:

```
msf6 auxiliary(scanner/winrm/winrm_login) > sessions

Active sessions
===============

  Id  Name  Type           Information                Connection
  --  ----  ----           -----------                ----------
  3         shell windows  WinRM a:p4$$w0rd (ADF2\a)  WinRM (192.168.123.15)

msf6 auxiliary(scanner/winrm/winrm_login) > sessions -i 3
[*] Starting interaction with 3...

Windows PowerShell 
Copyright (C) Microsoft Corporation. All rights reserved.

PS C:\Users\a> dir
dir


    Directory: C:\Users\a


Mode                LastWriteTime         Length Name                                                                  
----                -------------         ------ ----                                                                  
d-r---        8/17/2021   3:51 PM                3D Objects                                                            
... etc ...
```

### windows server 2012 Domain controller

Initially seemed to work:
```
msf6 auxiliary(scanner/winrm/winrm_login) > run

[!] No active DB -- Credential data will not be saved!
[+] 192.168.123.13:5985 - Login Successful: WORKSTATION\a:p4$$w0rd
[*] Command shell session 2 opened (WinRM) at 2021-09-07 18:25:35 +0100
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

However, it looks like interacting with the session seems to hang:
```
msf6 auxiliary(scanner/winrm/winrm_login) > sessions

Active sessions
===============

  Id  Name  Type           Information               Connection
  --  ----  ----           -----------               ----------
  2         shell windows  WinRM a:p4$$w0rd (ADF\a)  WinRM (192.168.123.13)

msf6 auxiliary(scanner/winrm/winrm_login) > sessions -i -1
[*] Starting interaction with 2...

Windows PowerShell
Copyright (C) 2014 Microsoft Corporation. All rights reserved.

PS C:\Users\a> dir

whoami
```

### Upgrading to meterpreter

Noticed that upgrading to Meterpreter seems to fail for both 2012 and 2019:

```
msf6 auxiliary(scanner/winrm/winrm_login) > sessions

Active sessions
===============

  Id  Name  Type           Information               Connection
  --  ----  ----           -----------               ----------
  1         shell windows  WinRM a:p4$$w0rd (ADF\a)  WinRM (192.168.123.13)

msf6 auxiliary(scanner/winrm/winrm_login) > sessions -u 1
[*] Executing 'post/multi/manage/shell_to_meterpreter' on session(s): [1]

[*] Upgrading session ID: 1
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 192.168.123.1:4433 
[-] Post failed: NoMethodError undefined method `get_once' for nil:NilClass
[-] Call stack:
[-]   /Users/user/Documents/code/metasploit-framework/lib/msf/base/sessions/command_shell.rb:665:in `shell_read'
[-]   /Users/user/Documents/code/metasploit-framework/lib/msf/core/session/provider/single_command_shell.rb:129:in `shell_command_token_win32'
[-]   /Users/user/Documents/code/metasploit-framework/lib/msf/core/session/provider/single_command_shell.rb:85:in `shell_command_token'
[-]   /Users/user/Documents/code/metasploit-framework/lib/msf/core/post/common.rb:124:in `cmd_exec'
[-]   /Users/user/Documents/code/metasploit-framework/lib/msf/core/post/windows/powershell.rb:42:in `have_powershell?'
[-]   /Users/user/Documents/code/metasploit-framework/modules/post/multi/manage/shell_to_meterpreter.rb:164:in `run'
```

File: lib/msf/util/signallable_event.rb [hunk 9]
```diff
@@ -0,0 +1,33 @@
+module Msf                                            
+module Util
+class SignallableEvent
```
[the remark this answers is not in the document]
Comment (smashery) [reply lib/msf/util/signallable_event.rb:3]: Thanks - that's exactly what I was looking for.

File: lib/net/winrm/connection.rb (hunk 1 above)
Comment (smashery) [reply lib/net/winrm/connection.rb:248]: I've made the change; however I'm not sure what you mean about an SSH transport. The intention of `connection_opts[:transport]` is to tell the WinRM library which Transport subclass to instantiate: Plaintext? Negotiate? Kerberos? CredSSP? We want none of these built-in transports, because they all use HTTPClient. Instead, we want our own class: `RexHttpTransport`. I don't think SSH is relevant here. Let me know if you think the code needs additional clarification in there.

File: lib/msf/base/sessions/winrm_command_shell.rb (hunk 2 above)
Comment (smashery) [reply lib/msf/base/sessions/winrm_command_shell.rb]: Done - currently just using the same connection across requests.

File: lib/msf/core/exploit/remote/winrm.rb (hunk 3 above)
Comment (smashery) [reply lib/msf/core/exploit/remote/winrm.rb]: Yep, typo - thanks.

File: lib/msf/core/exploit/remote/winrm.rb (hunk 4 above)
Comment (smashery) [reply lib/msf/core/exploit/remote/winrm.rb]: Debugging code left in by accident - removed.

File: lib/msf/base/sessions/winrm_command_shell.rb (hunk 6 above)
Comment (smashery) [reply lib/msf/base/sessions/winrm_command_shell.rb]: Done.

File: lib/msf/core/session_manager.rb (hunk 7 above)
Comment (smashery) [reply lib/msf/core/session_manager.rb]: True - makes sense.

**Comment by smashery** [comment]
Thanks @adfoster-r7 - I've resolved the issue with meterpreter upgrading. I'm not sure what's going on with 2012. I'm seeing it sending "OperationTimeout" messages every time I send stdin to it; will have to investigate that one further.

**Comment by smashery** [comment]
So the issue with 2012 is twofold:
- powershell and cmd both need `\r\n` at the end of a line to recognise it as a command; from 2012R2 onwards they seem to be fine with `\n` (which is what it was originally). So that lets you use cmd.exe and everything seems fine; but...
- PowerShell 3.0 (the default version on 2012) seems not to actually send anything to stdout or stderr, other than the initial banner; it just prints everything else to the PowerShell host. As we're just launching it from the command line and directing stdout/stderr to the winrm connection, we see nothing come back across the shell. I get the same behaviour if I launch it from meterpreter with `shell` and then run `powershell.exe`.

I've considered a few options:
1. Tried to find a way to direct the PowerShell success/error streams to stdout/stderr in PS <=3.0; couldn't see any way of doing so. If anyone knows of one, that'd be swell.
1. WinRM does support running powershell as a host (powershell remoting), which should give the streams we want, but I've read that that was introduced in Win7/Server 2008 R2, in which case it wouldn't work on Vista/2008.
1. Or we could keep it with the approach of launching a process manually, using, stdin but using cmd.exe rather than powershell.exe. Users can then choose to run powershell.exe within it if they want (it'll run but give no input on PS <= 3.0); or we could make it a configurable option (default cmd.exe), or put some smarts into auto-selecting.

I think 3 is the go, unless anyone knows of a way to make 1 work, or if I'm wrong, and Vista/2008 does support Powershell Remoting (I'll verify this tomorrow).

File: lib/net/winrm/connection.rb [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'winrm'
+require 'winrm/wsmv/write_stdin'
+
+module Net
+  module MsfWinRM
+    class RexWinRMConnection < WinRM::Connection
+      class MessageFactory < WinRM::PSRP::MessageFactory
```
Comment (adfoster-r7) [inline lib/net/winrm/connection.rb:12]: Just confirming, is this class needed? It doesn't look like we're dealing with the PSRP protocol here

File: lib/net/winrm/connection.rb [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+              # http://msdn.microsoft.com/en-us/library/cc251676.aspx
+              if err.fault_code == '2150858793'
+                yield nil, nil
+              else
+                raise
+              end
+            end
+          end
+    
+          def send_stdin(input)
```
Comment (adfoster-r7) [inline lib/net/winrm/connection.rb:120]: If we're providing our own overrides to `ReceiveResponseReader`, would it make sense to match the expected API in command shell and set rstream to be the `ReceiveResponseReader` object?

i.e. Updating this class to have a method for `write` and `get_once`. This would allow us to update `winrm_command_shell` to no longer override the base class functions, and use the normal implementation as defined here:

https://github.com/rapid7/metasploit-framework/blob/2f0509f2b56dbb73f28a19d5956de22c0bde0e91/lib/msf/base/sessions/command_shell.rb#L663-L673

File: lib/msf/base/sessions/winrm_command_shell.rb [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+      @buffer = []
+      if length > -1 and result.length > length
+        # Return up to length, and keep the rest in the buffer
+        extra = result[length..-1]
+        result = result[0,length]
+        @buffer << extra
+      end
+    }
+    result
+  end
```
Comment (adfoster-r7) [inline lib/msf/base/sessions/winrm_command_shell.rb:67]:
```suggestion
```

I think we could delete these method overrides if we implement this approach: https://github.com/rapid7/metasploit-framework/pull/15632/files#r705357961

File: modules/auxiliary/scanner/winrm/winrm_cmd.rb [hunk 13]
```diff
@@ -31,17 +28,48 @@ def initialize
...
+                ssl: ssl,
+                user: datastore['USERNAME'],
+                password: datastore['PASSWORD'],
+                transport: :rexhttp,
+                :no_ssl_peer_verification => true,
+                :operation_timeout => 0.5,
+                :timeout => 20,
+                :retry_limit => 1,
+                :realm => datastore['DOMAIN']
+            )
```
Comment (adfoster-r7) [inline modules/auxiliary/scanner/winrm/winrm_cmd.rb:52]: I think for Ruby 3 support we'd have to wrap this in an explicit hash
https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/

```suggestion
    conn = Net::MsfWinRM::RexWinRMConnection.new(
           {
                endpoint: endpoint,
                host: rhost,
                port: rport,
                uri: uri,
                ssl: ssl,
                user: datastore['USERNAME'],
                password: datastore['PASSWORD'],
                transport: :rexhttp,
                :no_ssl_peer_verification => true,
                :operation_timeout => 0.5,
                :timeout => 20,
                :retry_limit => 1,
                :realm => datastore['DOMAIN']
            }
     )
```

File: lib/net/winrm/connection.rb [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+            end
+          end
+
+          def localinfo
+            if self.http_client and self.http_client.conn
+                self.http_client.conn.localinfo
+            end
+          end
+
+          def ntlm_transform_response(ntlm_client, response)
```
Comment (adfoster-r7) [inline lib/net/winrm/connection.rb:194]: Just to ask the obvious, what are the reasons for overriding these methods? 👀

File: modules/auxiliary/scanner/winrm/winrm_cmd.rb [hunk 15]
```diff
@@ -31,17 +28,48 @@ def initialize
...
+      path = store_loot("winrm.cmd_results", "text/plain", ip, nil, "winrm_cmd_results.txt", "WinRM CMD Results")
+      f = File.open(path,'wb')
+      output = shell.run(datastore['CMD']) do |stdout,stderr|
+        stdout&.each_line do |line|
+          print_line(line.rstrip!)
+          f.puts(stdout)
+        end
+        print_error(stderr) if stderr
+      end
+      f.close
```
Comment (adfoster-r7) [inline modules/auxiliary/scanner/winrm/winrm_cmd.rb:65]: `File.open` can take a proc which should autoclose `f` for us:

```suggestion
      File.open(path,'wb') do |f|
        output = shell.run(datastore['CMD']) do |stdout,stderr|
          stdout&.each_line do |line|
            print_line(line.rstrip!)
            f.puts(stdout)
          end
          print_error(stderr) if stderr
        end
      end
```

File: modules/auxiliary/scanner/winrm/winrm_login.rb [hunk 16]
```diff
@@ -82,6 +110,41 @@ def run_host(ip)
 
   end
 
+def session_setup(shell,rhost,rport,endpoint)
```
Comment (adfoster-r7) [inline modules/auxiliary/scanner/winrm/winrm_login.rb:113]:
```suggestion
  def session_setup(shell,rhost,rport,endpoint)
```

**Comment by adfoster-r7** [comment]
We use Rubocop and msftidy to ensure the quality of our code. This can be ran from the root directory of Metasploit:

```
rubocop <directory or file>
tools/dev/msftidy.rb <directory or file>
```

You can automate most of these changes with the `-a` flag:

```
rubocop -a <directory or file>
```

I think we'd be good to run that on the new files that we're adding, as well as the winrm modules

File: modules/auxiliary/scanner/winrm/winrm_login.rb [hunk 17]
```diff
@@ -74,6 +75,33 @@ def run_host(ip)
...
+                      ssl: ssl,
+                      user: result.credential.public,
+                      password: result.credential.private,
+                      transport: :rexhttp,
+                      :no_ssl_peer_verification => true,
+                      :operation_timeout => 1, # For the WinRM server
+                      :timeout => 20, # For the underlying HTTP client
+                      :retry_delay => 1,
+                      :realm => result.credential.realm,
+                 )
```
Comment (adfoster-r7) [inline modules/auxiliary/scanner/winrm/winrm_login.rb:99]: https://github.com/rapid7/metasploit-framework/pull/15632/files#r705363789

File: lib/msf/util/signallable_event.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (smashery) [reply lib/msf/util/signallable_event.rb]: Deleted.

File: lib/net/winrm/connection.rb (hunk 10 above)
Comment (smashery) [reply lib/net/winrm/connection.rb]: Thanks - deleted.

File: lib/net/winrm/connection.rb (hunk 11 above)
Comment (smashery) [reply lib/net/winrm/connection.rb]: Agree that it'd be nice to have stream-like behaviour to minimise the code duplication; although `ReceiveResponseReader` probably isn't the correct place, as it's solely responsible for reading responses (not sending).

I split the stream-like behaviour of the command shell object into its own object (`WinRMStreamAdapter`) so we can let the base class use `rstream` in the cases you describe.

https://github.com/smashery/metasploit-framework/blob/winrm_improvements/lib/msf/base/sessions/winrm_command_shell.rb#L13-L144

File: lib/net/winrm/connection.rb (hunk 14 above)
Comment (smashery) [reply lib/net/winrm/connection.rb]: Good call; comments added. As far as I can tell from the content-type, I can't find anything else anywhere that uses this, other than WSMV, so didn't want to put it directly into the HttpClient itself.

File: lib/net/winrm/connection.rb [hunk 18]
```diff
@@ -0,0 +1,289 @@
...
+    
+            # Reads streams sent in one or more receive response messages
+            # @param wsmv_message [WinRM::WSMV::Base] A wsmv message to send to endpoint
+            # @param wait_for_done_state whether to poll for a CommandState of Done
+            # @yieldparam [Hash] Hash representation of stream with type and text
+            # @yieldparam [REXML::Document] Complete SOAP envelope returned to wsmv_message
+            def read_response(wsmv_message, wait_for_done_state = false)
+              resp_doc = nil
+              until command_done?(resp_doc, wait_for_done_state)
+                logger.debug('[WinRM] Waiting for output...')
```
[the remark this answers is not in the document]
Comment (smashery) [reply lib/net/winrm/connection.rb]: Added an adapter to MSF's logging:

https://github.com/smashery/metasploit-framework/blob/winrm_improvements/lib/net/winrm/connection.rb#L45-L61

File: modules/auxiliary/scanner/winrm/winrm_cmd.rb (hunk 15 above)
Comment (smashery) [reply modules/auxiliary/scanner/winrm/winrm_cmd.rb]: Now writing all at once at the end:

https://github.com/smashery/metasploit-framework/blob/winrm_improvements/modules/auxiliary/scanner/winrm/winrm_cmd.rb#L59-L68

File: modules/auxiliary/scanner/winrm/winrm_login.rb (hunk 17 above)
Comment (smashery) [reply modules/auxiliary/scanner/winrm/winrm_login.rb]: Now using a hash:

https://github.com/smashery/metasploit-framework/blob/winrm_improvements/modules/auxiliary/scanner/winrm/winrm_login.rb#L84-L98

File: modules/auxiliary/scanner/winrm/winrm_cmd.rb (hunk 13 above)
Comment (smashery) [reply modules/auxiliary/scanner/winrm/winrm_cmd.rb]: Change made - https://github.com/smashery/metasploit-framework/blob/winrm_improvements/modules/auxiliary/scanner/winrm/winrm_cmd.rb#L40-L54

**Comment by smashery** [comment]
Thanks for the review @adfoster-r7; really helpful. I'm still chasing down a bug in upgrading this kind of shell to meterp on 2012; root cause seems to be that the 2012 shell is not reflecting stdin back to stdout (which seems to be assumed for Windows shells).

**Comment by smcintyre-r7** [comment]
I'm getting an error when I try to run a command against multiple sessions with `sessions -c` that seems to come from the fact that the command shell session isn't a selectable IO.

```
msf6 auxiliary(scanner/winrm/winrm_login) > run USERNAME=smcintyre PASSWORD=Password1

[!] No active DB -- Credential data will not be saved!
[+] 192.168.159.42:5985 - Login Successful: WORKSTATION\smcintyre:Password1
[*] Command shell session 1 opened (192.168.159.29:45551 -> 192.168.159.42:5985) at 2021-09-20 17:25:23 -0400
[*] Scanned 1 of 2 hosts (50% complete)
[!] No active DB -- Credential data will not be saved!
[+] 192.168.159.96:5985 - Login Successful: WORKSTATION\smcintyre:Password1
[*] Command shell session 2 opened (192.168.159.29:44063 -> 192.168.159.96:5985) at 2021-09-20 17:25:24 -0400
[*] Scanned 2 of 2 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/winrm/winrm_login) > sessions -c "systeminfo"
[*] Running 'systeminfo' on shell session 1 (192.168.159.42)
[-] Session manipulation failed: no implicit conversion of nil into IO ["/home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/base/sessions/command_shell.rb:649:in `select'", "/home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/base/sessions/command_shell.rb:649:in `shell_command'", "/home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/ui/console/command_dispatcher/core.rb:1444:in `block (2 levels) in cmd_sessions'", "/home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/ui/console/command_dispatcher/core.rb:1413:in `each'", "/home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/ui/console/command_dispatcher/core.rb:1413:in `block in cmd_sessions'", "/home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/ui/console/command_dispatcher/core.rb:1403:in `each'", "/home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/ui/console/command_dispatcher/core.rb:1403:in `cmd_sessions'", "/home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/ui/text/dispatcher_shell.rb:555:in `run_command'", "/home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/ui/text/dispatcher_shell.rb:504:in `block in run_single'", "/home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/ui/text/dispatcher_shell.rb:498:in `each'", "/home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/ui/text/dispatcher_shell.rb:498:in `run_single'", "/home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/ui/text/shell.rb:162:in `run'", "/home/smcintyre/Repositories/metasploit-framework.pr/lib/metasploit/framework/command/console.rb:48:in `start'", "/home/smcintyre/Repositories/metasploit-framework.pr/lib/metasploit/framework/command/base.rb:82:in `start'", "./msfconsole:23:in `<main>'"]
msf6 auxiliary(scanner/winrm/winrm_login) >
```

To reproduce, obtain one or more WinRM sessions and then run `sessions -c "systeminfo"`. Our standard shell sessions will execute it as expected.

The other test I ran was using a post module and it worked as expected:
```
msf6 auxiliary(scanner/winrm/winrm_login) > run

[!] No active DB -- Credential data will not be saved!
[+] 192.168.159.42:5985 - Login Successful: WORKSTATION\smcintyre:Password1
[*] Command shell session 5 opened (192.168.159.29:32787 -> 192.168.159.42:5985) at 2021-09-20 17:30:54 -0400
[*] Scanned 1 of 2 hosts (50% complete)
[!] No active DB -- Credential data will not be saved!
[+] 192.168.159.96:5985 - Login Successful: WORKSTATION\smcintyre:Password1
[*] Command shell session 6 opened (192.168.159.29:45845 -> 192.168.159.96:5985) at 2021-09-20 17:30:55 -0400
[*] Scanned 2 of 2 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/winrm/winrm_login) > use post/windows/gather/checkvm 
msf6 post(windows/gather/checkvm) > set SESSION 5
SESSION => 5
msf6 post(windows/gather/checkvm) > run

[*] Checking if the target is a Virtual Machine ...
[+] This is a VMware Virtual Machine
[*] Post module execution completed
msf6 post(windows/gather/checkvm) > set SESSION 6
SESSION => 6
msf6 post(windows/gather/checkvm) > run

[*] Checking if the target is a Virtual Machine ...
[+] This is a VMware Virtual Machine
[*] Post module execution completed
msf6 post(windows/gather/checkvm) >
```

**Comment by smcintyre-r7** [comment]
# Release Notes
This improves Metasploit's WinRM capabilities by allowing shell sessions to be established over the protocol. The shell sessions are interactive and able to be used with post modules.
